### PR TITLE
CLOUDP-126253: Add network peering feature

### PR DESCRIPTION
**Pull request by Sugar-pack**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: pkg/api/v1/status/atlasproject.go [hunk 1]
```diff
@@ -34,6 +34,14 @@ func AtlasProjectSetPrivateEnpointsOption(privateEndpoints []ProjectPrivateEndpo
 	}
 }
 
+func AtlasProjectSetNetworkPeerOption(networkPeers *[]AtlasNetworkPeer) AtlasProjectStatusOption {
+	return func(s *AtlasProjectStatus) {
+		if networkPeers != nil {
```
Comment (fabritsius) [inline pkg/api/v1/status/atlasproject.go:39]: Here it's probably better to remove this line for cases where we actually want to reset the status. Like when someone decided to remove Network Peering we don't want status to stay frozen in the last state we had it enabled.

File: test/e2e/actions/networkpeer/gcp.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		ExchangeSubnetRoutes: true,
+	}
+	request := &compute.NetworksAddPeeringRequest{
+		NetworkPeering: peer,
+	}
+	_, err = networkService.AddPeering(gpcProjectID, vnetName, request).Do()
+	return err
+}
+
+func DeleteGCPvpc(gcpProjectID, vnetName string) error {
```
Comment (fabritsius) [inline test/e2e/actions/networkpeer/gcp.go:45]: This function name seems strange as `vpc` is lowercased. Also might be a good idea to have some standard naming for these functions for search purposes (like `DeleteVPCforAWS`, `DeleteVPCforAzure`, `DeleteVPCforGCP`). Not necessary ofc but what do you think?

File: test/e2e/actions/networkpeer/networkpeer.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+				errList = append(errList, err)
+			}
+		}
+	}
+	return errList
+}
+
+func EstablishPeerConnections(peers []status.AtlasNetworkPeer) error {
+	for _, peerStatus := range peers {
+		switch peerStatus.ProviderName {
```
Comment (fabritsius) [inline test/e2e/actions/networkpeer/networkpeer.go:75]: I assume `Azure` will also be added at some point :)

File: pkg/controller/atlasproject/atlasproject_controller.go [hunk 4]
```diff
@@ -160,6 +160,10 @@ func (r *AtlasProjectReconciler) Reconcile(context context.Context, req ctrl.Req
 					setCondition(ctx, status.PrivateEndpointReadyType, result)
 					return result.ReconcileResult(), nil
 				}
+				if result = DeleteAllNetworkPeers(context, projectID, ctx.Client.Peers, ctx.Log); !result.IsOk() {
+					setCondition(ctx, status.NetworkPeerReadyType, result)
```
Comment (fabritsius) [inline pkg/controller/atlasproject/atlasproject_controller.go:164]: When deleting it also makes sense to unset the conditions for the peers 🤔

**Review by fabritsius — CHANGES_REQUESTED** [review]
Looks great 👍 just have a couple of small suggestions :)

File: pkg/api/v1/status/atlasproject.go (hunk 1 above)
Comment (Sugar-pack) [reply pkg/api/v1/status/atlasproject.go]: Done

File: test/e2e/actions/networkpeer/gcp.go (hunk 2 above)
Comment (Sugar-pack) [reply test/e2e/actions/networkpeer/gcp.go]: Done

File: test/e2e/actions/networkpeer/networkpeer.go (hunk 3 above)
Comment (Sugar-pack) [reply test/e2e/actions/networkpeer/networkpeer.go]: Not really. Azure has a different approach. For this case, I added a comment.

File: pkg/controller/atlasproject/atlasproject_controller.go (hunk 4 above)
Comment (Sugar-pack) [reply pkg/controller/atlasproject/atlasproject_controller.go:164]: What do you mean? When the project deleting, why is it makes sense to unset the conditions for the peers?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -7,4 +7,4 @@ repos:
   - repo: https://github.com/shellcheck-py/shellcheck-py
     rev: v0.7.1.1
     hooks:
-      - id: shellcheck
+      - id: shellcheck
```
Comment (fabritsius) [inline .pre-commit-config.yaml:10]: Revert :)

File: pkg/controller/atlasproject/atlasproject_controller.go (hunk 4 above)
Comment (fabritsius) [reply pkg/controller/atlasproject/atlasproject_controller.go:164]: Ah, true, no reason to do it here, I mixed the spot but it still makes sense to unset them somewhere)

**Review by fabritsius — APPROVED** [review]
Looks Great 👍

File: test/e2e/actions/networkpeer/networkpeer.go (hunk 3 above)
Comment (fabritsius) [reply test/e2e/actions/networkpeer/networkpeer.go]: Ok 👌

File: pkg/controller/atlasproject/atlasproject_controller.go (hunk 4 above)
Comment (Sugar-pack) [reply pkg/controller/atlasproject/atlasproject_controller.go:164]: Sure. Please look at https://github.com/mongodb/mongodb-atlas-kubernetes/blob/dc943a1cf0ab10859525f6b2d067ee0e4522cdae/pkg/controller/atlasproject/network_peering.go#L43

File: .pre-commit-config.yaml (hunk 5 above)
Comment (Sugar-pack) [reply .pre-commit-config.yaml:10]: May you explain do you mean?

File: pkg/controller/atlasproject/atlasproject_controller.go (hunk 4 above)
Comment (fabritsius) [reply pkg/controller/atlasproject/atlasproject_controller.go:164]: Great 👍

File: .pre-commit-config.yaml (hunk 5 above)
Comment (fabritsius) [reply .pre-commit-config.yaml]: For some reason it doesn't show it here but the only difference between these lines is that new one doesn't have new line symbol at the end. This change is completely unnecessary.

**Pull request merged**